app_engine/main.py

```diff
@@ -0,0 +1,54 @@
+from flask import Flask, jsonify, request
+from flask_cors import CORS
+import redis_utils
+import requests
+import json
+
+
+app = Flask(__name__)
+cors = CORS(
+    app,
+    resources={
+        r"/search/*": {"origin": "*"},
+        r"/test/*": {"origin": "*"},
+    },
+)
+
+redis_connection = redis_utils.RedisConnect()
+client = redis_connection.get_client()
+
+
+@app.route("/search", methods=["GET"])
+def get_images():
+    tag = request.args.get("t")
+    query = request.args.get("s_query")
+    top_k = request.args.get("k")
+
+    if client.get(tag):
+        result = json.loads(client.get(tag))
+        top_urls = result["top_urls"]
+        top_scores = result["top_scores"]
+
+    else:
```

Comment:
I am not 100% sure about this lines of code. 
But shouldn't this condition be made based on `query`? `tag` is for PIXABAY searching keywords.

So, whenever a user inputs the same tag like `Tigers` when the query is whether `Tigers on mountain` or `Tigers on snow`, the result is going to be the same. If this behaviour is correct, I think this is problematic. What do you think?

Reply:
You are absolutely right. Getting back to you in the Google Doc.

Reply:
@deep-diver regarding this part (cached semantic queries), what all should we first cache, and what should be the structure of a single entry? 

Should it be semantic queries as the keys and the lists of top URLs as the values?

Reply:
hmm... I would go with the "Should it be semantic queries as the keys and the lists of top URLs as the values" as a pretty simple baseline!

Reply:
Alright. Will update here once done.

Reply:
While are we at it could you look into connecting MemoryStore with our API hosted via GCE VM? I think if we are able to do that then we can bypass the App Engine workflow and improve the latency.

Reply:
It is possible to get rid of the app engine and bypass the request to the VM! We can try since VM can get access to the MemoryStore :) 

My final goal is to move raw VM solution to GKE solution after all! But it all depends on what we want to achieve and the end goal of this project. 

When do you think we can report the work to Soonson for asking writing a blog post?

Reply:
> It is possible to get rid of the app engine and bypass the request to the VM! We can try since VM can get access to the MemoryStore :)

Yeah, that's why. 

> y final goal is to move raw VM solution to GKE solution after all! But it all depends on what we want to achieve and the end goal of this project.

I think that would be actually better since it'd improve the maturity of the project. This (improving maturity) led me to incorporate App Engine and Caching in the first place actually. 

> When do you think we can report the work to Soonson for asking writing a blog post?

How about we start putting together the draft briefly and then reach out to him and Karl?

Reply:
Also, just addressed [your comment](https://github.com/deep-diver/image_search_with_natural_language/pull/14#discussion_r659275104).

Reply:
> Additionally, for the end goal, I think once we are done incorporating caching capabilities in our raw VM and shipping it as a GKE solution, the project can be deemed complete. WDYT?

Sure! That could be the end!
Just thinking about the reader's perspective like what they are expecting when reading our article ;) 

I am going to merge your PR. let's talk more at the Google Docs then

Reply:
Sure. As soon as we have a mutual agreement in when we should let Soonson know I can send an email. 

Honestly, I think we are in a good position to do that now as well. Let me know. But no matter what their decision is I think we should definitely write a blog post about the project :)